### PR TITLE
Development

### PR DESCRIPTION
**Pull request by shandralor**
Added remove user friends functionality. Not sure what needs to happen next, but I think you need to merge some conflicts @Lukas-Batema

### CODE REVIEW

**Comment by shandralor** [comment]
Please review the option 7 push also. It implements viewing message titles only and being able to choose an individual message to view. If you view it, the rating increase function will get triggered.

File: server/db.py [hunk 1]
```diff
@@ -334,3 +375,72 @@ def get_encrypted_sym_key(username: str, user_password, friend_username:str):
...
+                "text": message,
+                "date_time": str(datetime.datetime.utcnow())
+            }    }  }
+        try:
+            DM_MESSAGES.put(update)
+        except Exception as error_message:
+            logging.exception(error_message)
+            return
+
+update_thought_rating("222a8c6f-13a0-4f4c-8ae9-8f1e2dae6b8a")
```
Comment (Lukas-Batema) [inline server/db.py:446]: What is this? I'd put this in an environment variable if I were you.

**Review by luni-moon — COMMENTED** [review]
I am gonna test this locally, just to be sure, as this is a big change ;)

File: client/cli_main.py [hunk 2]
```diff
@@ -230,6 +266,72 @@ def main():
...
+                        print()
+                        for friend in friends:
+                            print(f"- {friend}")
+                            print()
+                        friend_username = input("Enter the username of the friend you want to remove:")
+                        print()
+                        remove_user_friends(server_url, friend_username)
+                        #reloading friends object after removing a friend
+                        friends = get_user_friends(server_url)
+                    elif sub_choice == "7":
```
Comment (Lukas-Batema) [inline client/cli_main.py:281]: Currently this is giving me this (error at the bottom):
```txt
PS D:\Programming\tmp\PeerBrain\client> python cli_main.py

Welcome to version 0.2 of the Brainwaves P2P client!
-------------------------------------------------------------------
The currently set server url is https://peerbrain.teckhawk.be/
-------------------------------------------------------------------

Please select what you want to do from the menu below.
1. Log in to the current server
2. Register account on the current server
3. Change server
4. Reset your password
Q to exit
>> 1

Logged in with valid token


MAIN MENU:


Please choose an option:

1. TECHNICAL ACTIONS
2. ACCOUNT ACTIONS
3. LOG OUT
Q to exit
>> 2

ACCOUNT MENU:


Please choose an option:

1. Check your account details
--------------------------------
2. Create a message
3. Show all messages from a friend
--------------------------------
4. Add a friend
5. Check friends list
6. Remove a friend from your friend list
--------------------------------
7. New thought display function test
B to return to main menu
>> 7

Please confirm your password to get your messages:


Please enter the username of the friend that you want to see messages from:

admin
Please choose a thought to read by entering its number or type B to go back:

1. TITLE: testing as admin       RATING: 1.2375035237499352
2. TITLE: testing message for nishant    RATING: 1.2375035237499352
3. TITLE: testing message while refactor cli     RATING: 2.1724833783268154
4. TITLE: The meaning of life?   RATING: 1.2375035237499352

Enter thought number: 4

TITLE:  The meaning of life?
RATING: 1.2375035237499352

MESSAGE:  You thought there would be something here, didn't you :-P.


Traceback (most recent call last):
  File "D:\Programming\tmp\PeerBrain\client\cli_main.py", line 360, in <module>
    main()
  File "D:\Programming\tmp\PeerBrain\client\cli_main.py", line 325, in main
    update_rating_for_thought(server_url, str(selected_thought["key"]))
KeyError: 'key'
NativeCommandExitException: Program "python.exe" ended with non-zero exit code: 1.
```

**Review by luni-moon — CHANGES_REQUESTED** [review]
Currently, I'm getting this error on option 7, will fix tonight!

File: client/cli_main.py [hunk 3]
```diff
@@ -306,35 +307,38 @@ def main():
...
-                                            except FileNotFoundError as err:
-                                                print("Error decrypting message, you may need to generate your keys still!\nError:", err)
-                                            except ValueError as err:
-                                                print("Please restart the programme to register your keys!\nError:", err)
+                                            decrypted_message = decrypt_message(selected_thought['content'], encrypted_sym_key)
+                                            print(f"MESSAGE:  {decrypted_message}\n\n")
+                                            key_path = os.path.join(os.path.dirname(__file__), 'keys', 'message.key')
+                                            with open(key_path, 'rb') as key_file:
+                                                key = key_file.readline(1)
+                                            update_rating_for_thought(server_url, key)
```
Comment (Lukas-Batema) [inline client/cli_main.py:328]: I'm not sure the key is supposed to be a str type, so I changed it to bytes, but it still is throwing a 404 error...

File: client/client_functions.py [hunk 4]
```diff
@@ -176,19 +176,19 @@ def get_user_friends(server_url:str)->tuple:
...
     """function to update a thoughts rating when it is read"""
     account_url_suffix = "api/v1/update-thought-rating"
 
     headers = {"Authorization": f"Bearer {get_token()}"}
 
-    params = {"key" : key}
+    params = {"key": key}
     
-    response = requests.get(f"{server_url}{account_url_suffix}", headers=headers, params = params, timeout=10)
+    response = requests.get(f"{server_url}{account_url_suffix}", headers=headers, params=params["key"], timeout=10)
```
Comment (Lukas-Batema) [inline client/client_functions.py:187]: Had to update this line in accordance to the return type screams

**Review by luni-moon — CHANGES_REQUESTED** [review]
Still needs to be fixed, so I'm requesting changes

File: server/db.py (hunk 1 above)
Comment (shandralor) [reply server/db.py:446]: it is a dynamically generated hash based on the two usersnames. It will be used to easily retrieve conversations. If you know the username and the friends username the correct hash will be generated(they are sorted alphabetically and trimmed to 5 characters if longer than 5, before hash generation). It is not so much a security feature. It is more to easily store and retrieve the conversations. There will be one such hash per conversation between two people

File: client/cli_main.py (hunk 2 above)
Comment (shandralor) [reply client/cli_main.py:282]: It is already fixed in the server code. The get thoughts function did not return the key initially. I've made changes that do return it, but you need to run the dev server. You are testing this on main, which does not have the altered server code yet.

File: client/cli_main.py (hunk 3 above)
Comment (shandralor) [reply client/cli_main.py:328]: it is a string type normally, see my previous answer :-)

File: client/client_functions.py (hunk 4 above)
Comment (shandralor) [reply client/client_functions.py:187]: that is kinda ridiculous lol... I wasn't even aware this was a thing in python

File: server/db.py (hunk 1 above)
Comment (shandralor) [reply server/db.py:446]: plus that update_thought_rating at the bottom needs to go, i was testing functionality :-P

File: client/cli_main.py (hunk 2 above)
Comment (Lukas-Batema) [reply client/cli_main.py:285]: My bad, will test on dev

File: server/db.py (hunk 1 above)
Comment (Lukas-Batema) [reply server/db.py]: Understood

File: client/cli_main.py [hunk 5]
```diff
@@ -163,7 +188,10 @@ def main():
                     print("--------------------------------")
                     print("4. Add a friend")
                     print("5. Check friends list")
+                    print("6. Remove a friend from your friend list")
                     print("--------------------------------")
+                    print("7. New thought display function test")
```
Comment (Lukas-Batema) [inline client/cli_main.py:193]: You may want to change this name to something better now.

File: client/cli_main.py [hunk 6]
```diff
@@ -228,11 +268,91 @@ def main():
...
+                                                reading = False
+                                    except FileNotFoundError as err:
+                                        print("Error decrypting message, you may need to generate your keys still!\nError:", err)
+                                    except ValueError as err:
+                                        print("Please restart the programme to register your keys!\nError:", err)
+                                    except IndexError as err:
+                                        print("You selected a non-existant thought number.\nError:", err)
+                    elif sub_choice == "8":
+                        friend_username = input("Please enter your friends username: \n")
+                        get_user_conversation(server_url, friend_username)
```
Comment (Lukas-Batema) [inline client/cli_main.py:343]: I'm getting this error (whole output, error near bottom (404 doubled):
```txt
@Lukas-Batema ➜ /workspaces/PeerBrain (development) $ python client/cli_main.py

Welcome to version 0.2 of the Brainwaves P2P client!
-------------------------------------------------------------------
The currently set server url is https://peerbrain.teckhawk.be/
-------------------------------------------------------------------

Please select what you want to do from the menu below.
1. Log in to the current server
2. Register account on the current server
3. Change server
4. Reset your password
Q to exit
>> 1

Logged in with valid token


MAIN MENU:


Please choose an option:

1. TECHNICAL ACTIONS
2. ACCOUNT ACTIONS
3. LOG OUT
Q to exit
>> 2

ACCOUNT MENU:


Please choose an option:

1. Check your account details
--------------------------------
2. Create a message
3. Show all messages from a friend
--------------------------------
4. Add a friend
5. Check friends list
6. Remove a friend from your friend list
--------------------------------
7. New thought display function test
8. Private conversations(DM)
B to return to main menu
>> 8
Please enter your friends username: 
admin



--------
{'detail': 'Not Found'}
--------
{'detail': 'Not Found'}
Error: 404

1. Yes
2. No
Do you want to reply to admin? 1
Enter your message: joe
Something went wrong with sending your message!


ACCOUNT MENU:


Please choose an option:

1. Check your account details
--------------------------------
2. Create a message
3. Show all messages from a friend
--------------------------------
4. Add a friend
5. Check friends list
6. Remove a friend from your friend list
--------------------------------
7. New thought display function test
8. Private conversations(DM)
B to return to main menu
>> b
Returning to main menu...

MAIN MENU:


Please choose an option:

1. TECHNICAL ACTIONS
2. ACCOUNT ACTIONS
3. LOG OUT
Q to exit
>> q
Goodbye!
```
Will that be fixed with the dev database that I don't have access to?

**Review by luni-moon — CHANGES_REQUESTED** [review]
Almost good. These still need to be changed.

File: client/cli_main.py (hunk 6 above)
Comment (Lukas-Batema) [reply client/cli_main.py:343]: Okay, it should just need these changes (basically renaming the option 7 of main option 2)

Comment (shandralor) [reply client/cli_main.py:343]: done :-)

File: client/cli_main.py [hunk 7]
```diff
@@ -163,7 +188,10 @@ def main():
                     print("--------------------------------")
                     print("4. Add a friend")
                     print("5. Check friends list")
+                    print("6. Remove a friend from your friend list")
                     print("--------------------------------")
+                    print("7. New system to view messages individually. This will also increase their rating.")
```
Comment (Lukas-Batema) [inline client/cli_main.py:193]: I'd reword it like so, "View your messages singularly".

**Review by luni-moon — APPROVED** [review]
Looks good!

**EDIT**: Please refer to the new review...

**Review by luni-moon — CHANGES_REQUESTED** [review]
Meant to request the last review as a change, wasn't thinking haha.

**Comment by shandralor** [comment]
done
